service/dynamodb/marshal.go

```diff
@@ -0,0 +1,233 @@
+package dynamodb
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+	"runtime"
+	"strconv"
+)
+
+// Marshal accepts a map or struct converts it to the map[string]*AttributeValue
+// type used to interact with the DynamoDB Item APIs.
+//
+// If in is a struct, we first JSON encode/decode it to get the data as a map.
+// This can/should be optimized later.
+func Marshal(in interface{}) (item map[string]*AttributeValue, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			if e, ok := r.(runtime.Error); ok {
+				err = e
+			} else if s, ok := r.(string); ok {
+				err = errors.New(s)
+			} else {
+				err = r.(error)
+			}
+			item = nil
+		}
+	}()
+
+	v := reflect.ValueOf(in)
+	switch v.Kind() {
+	case reflect.Struct:
+		item = marshalStruct(in)
+	case reflect.Map:
+		if v.Type().Key().Kind() != reflect.String {
+			return nil, errors.New("item must be a map[string]interface{} or struct (or a non-nil pointer to one), got " + v.Type().String())
+		}
+		item = marshalMap(in)
+	case reflect.Ptr:
+		if v.IsNil() {
+			return nil, errors.New("item must not be nil")
+		}
+		return Marshal(v.Elem().Interface())
+	default:
+		return nil, errors.New("item must be a map[string]interface{} or struct (or a non-nil pointer to one), got " + v.Type().String())
+	}
+	return item, nil
+}
+
+// Unmarshal accepts the map[string]*AttributeValue type returned by the
+// DynamoDB Item APIs and converts it to a map or struct.
+//
+// If v points to a struct, we first convert it to a basic map, then JSON
+// encode/decode it to convert to a struct. This can/should be optimized later.
+func Unmarshal(item map[string]*AttributeValue, v interface{}) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			if e, ok := r.(runtime.Error); ok {
+				err = e
+			} else if s, ok := r.(string); ok {
+				err = errors.New(s)
+			} else {
+				err = r.(error)
+			}
+			item = nil
+		}
+	}()
+
+	m := make(map[string]interface{})
+	for k, v := range item {
+		m[k] = unmarshal(v)
+	}
+
+	rv := reflect.ValueOf(v)
+	if rv.Kind() != reflect.Ptr || rv.IsNil() {
+		return errors.New("v must be a non-nil pointer to a map[string]interface{} or struct, got " + rv.Type().String())
+	}
+
+	switch rv.Elem().Kind() {
```

Comment:
That's true. The json encoding has always made me feel a little icky, but I haven't spent time to remove it because it just works.

Also, I haven't benchmarked it, but one nice thing about using json to convert is that it caches the reflection results. Though I guess in practice this isn't likely to be a hot spot either way, the data types are usually pretty small.